lib/common/socket/evloop.c.h

```diff
@@ -115,19 +115,97 @@ static void link_to_statechanged(struct st_h2o_evloop_socket_t *sock)
     }
 }
 
-static const char *on_read_core(int fd, h2o_buffer_t **input)
+#define NSEC_PER_SEC 1000000000
+
+static inline int64_t
+timespec_to_nsec(const struct timespec *a)
+{
+       return (int64_t)a->tv_sec * NSEC_PER_SEC + a->tv_nsec;
+}
+
+#if defined(__linux__) && !H2O_USE_LIBUV
+static void handle_timestamp(struct st_h2o_evloop_socket_t *sock, struct msghdr *msg)
+{
+    struct scm_timestamping *ts = NULL;
+    uint64_t packet_ts = 0;
+    h2o_loop_t *loop = sock->loop;
+
+    for (struct cmsghdr *cmsg = CMSG_FIRSTHDR(msg); cmsg;
+            cmsg = CMSG_NXTHDR(msg, cmsg)) {
+
+        /* in the future we may want to harvest IP_RECVERR messages here, which could have interesting data */
+        if (cmsg->cmsg_level != SOL_SOCKET)
+            continue;
+
+        switch (cmsg->cmsg_type) {
+            case SO_TIMESTAMPNS:
+                ts = (struct scm_timestamping *)CMSG_DATA(cmsg);
+                packet_ts = timespec_to_nsec(&ts->ts[0]);
+
+                h2o_sliding_counter_start(&loop->packet_latency_nanosec_counter, packet_ts);
```

Comment:
I'm not sure if I follow the logic here. IIUC, SO_TIMESTAMPNS is used to collect the latency between the NIC (or the driver) and userspace. But I do not see us using a timestamp collected within the userspace.

Reply:
err, sorry about that! i think i made an error when moving code between branches. I have fixed this to obtain the userland timestamp for generating the sliding window.